### ISSUE TITLE
The model saved after AWQ quantization is not quantized.

### ISSUE THREAD

**Issue by Minami-su** [vllm-project/llm-compressor#1533]
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from typing import Any, List, Dict # <--- CORRECTED IMPORT
from llmcompressor import oneshot
from llmcompressor.modifiers.awq import AWQModifier
from llmcompressor.modifiers.quantization import QuantizationModifier
from llmcompressor.modifiers.smoothquant import SmoothQuantModifier
import argparse
from compressed_tensors.quantization import (
    QuantizationArgs,
    QuantizationScheme,
    QuantizationStrategy,
    QuantizationType,
)
import lm_eval
from lm_eval.utils import make_table
##CUDA_VISIBLE_DEVICES=7 python /data/jcxy/haolu/workspace/training/alignment-handbook/scripts/llm_compressor.py
parser = argparse.ArgumentParser()
parser.add_argument("--model_path", default="/data2/jcxy/llm_model/Qwen3-0.6B", type=str)
parser.add_argument("--quant_path", default="/data2/jcxy/llm_model/Qwen3-0.6B-awq-test", type=str)
parser.add_argument("--data_path", default="/data/jcxy/haolu/workspace/store/train_data/ins_data/PsyLLM_InstructV6_250529_NR_DPO.jsonl", type=str)
args = parser.parse_args()
model_path = args.model_path
print(model_path)
quant_path = args.quant_path
print(quant_path)
data_path = args.data_path
print(data_path)
# Select number of samples. 256 samples is a good place to start.
# Increasing the number of samples can improve accuracy.
NUM_CALIBRATION_SAMPLES = 256
MAX_SEQUENCE_LENGTH = 512
N_PARALLEL_SAMPLES = 32
SAVE_DIR = quant_path

tokenizer = AutoTokenizer.from_pretrained(model_path)
tokenizer.add_special_tokens({"bos_token": "<|im_start|>"})
tokenizer.add_special_tokens({"eos_token": "<|im_end|>"})
tokenizer.add_special_tokens({"pad_token": "<|endoftext|>"})
data = load_dataset("json", data_files=data_path)
column_names = list(data["train"].features)

def build_tokenized_answer_with_labels(instruction: List[Dict], output: List[Dict]) -> Dict[str, Any]:
    """
    Tokenizes the instruction and output using apply_chat_template for the full conversation.
    Correctly handles multi-turn conversations with special template logic (e.g., <think> tags).
    Labels are set to -100 for tokens corresponding to role='system' and role='user'
    messages across the entire conversation.

    Args:
        instruction (List[Dict]): List of initial instruction messages (e.g., system, user).
        output (List[Dict]): List of subsequent messages, potentially multi-turn
                            (e.g., assistant, user, assistant, user, ...).

    Returns:
        Dict: A dictionary containing 'input_ids', 'attention_mask', and 'labels'.
    """
    full_conversation = instruction + output
    
    full_text = tokenizer.apply_chat_template(
        full_conversation,
        tokenize=False,
        add_generation_prompt=False,
        enable_thinking=True
    )
    
    tokenized_full = tokenizer(
        full_text,
        add_special_tokens=False,
        return_attention_mask=True
    )
    full_input_ids = tokenized_full["input_ids"]
    content_to_tokens = []
    current_position = 0
    
    for message in full_conversation:
        message_text = tokenizer.apply_chat_template(
            [message],
            tokenize=False,
            add_generation_prompt=False,
            enable_thinking=False  # Disable thinking for individual messages
        )
        
        message_tokens = tokenizer(
            message_text,
            add_special_tokens=False
        )["input_ids"]
        
        found = False
        for start_idx in range(current_position, len(full_input_ids) - len(message_tokens) + 1):
            if full_input_ids[start_idx:start_idx + len(message_tokens)] == message_tokens:
                content_to_tokens.append({
                    "role": message["role"],
                    "start": start_idx,
                    "end": start_idx + len(message_tokens)
                })
                current_position = start_idx + len(message_tokens)
                found = True
                break
        
        if not found:
            content_to_tokens.append({
                "role": message["role"],
                "start": current_position,
                "end": len(full_input_ids)
            })
            break
    labels = [-100] * len(full_input_ids)  # Initialize all to -100
    for item in content_to_tokens:
        if item["role"] == "assistant":
            # Set labels for assistant messages
            for idx in range(item["start"], item["end"]):
                labels[idx] = full_input_ids[idx]
    think_block_string = "<think>\n\n</think>\n\n" # Adjust this if your template differs
    think_block_tokens = tokenizer(think_block_string, add_special_tokens=False)["input_ids"]
    think_block_len = len(think_block_tokens)
    full_len = len(full_input_ids)
    i = 0
    while i <= full_len - think_block_len:
        if full_input_ids[i : i + think_block_len] == think_block_tokens:
            for j in range(i, i + think_block_len):
                labels[j] = -100 # Set label to -100
            i += think_block_len
        else:
            i += 1
    think_block_string = "<|im_start|>assistant\n" # Adjust this if your template differs
    think_block_tokens = tokenizer(think_block_string, add_special_tokens=False)["input_ids"]
    think_block_len = len(think_block_tokens)
    full_len = len(full_input_ids)
    i = 0
    while i <= full_len - think_block_len:
        if full_input_ids[i : i + think_block_len] == think_block_tokens:
            for j in range(i, i + think_block_len):
                labels[j] = -100 # Set label to -100
            i += think_block_len
        else:
            i += 1
    sum_text = tokenizer.decode(full_input_ids, skip_special_tokens=False)
    decoded_labels = [
            token_id if label != -100 else tokenizer.pad_token_id
            for token_id, label in zip(full_input_ids, labels)
        ]
    sum_labels = tokenizer.decode(decoded_labels, skip_special_tokens=False)
    return {'text': sum_text,
            "labels": sum_labels}

def generate_and_tokenize_prompt(data_point):
    """
    生成并令牌化提示信息，同时处理选择答案和拒绝答案，并计算长度。
    
    参数:
        data_point (Dict): 包含 'prompt', 'chosen' 和 'rejected' 的单个数据样本。
    
    返回:
        Dict: 包含 tokenized 输入和标签以及长度信息的字典。
    """
    prompt = data_point['prompt']    # 提示信息
    chosen = data_point['chosen']    # 选择的答案
    rejected = data_point['rejected']  # 拒绝的答案
    

    # 令牌化 chosen 和 rejected 答案并生成标签
    dicts_chosen = build_tokenized_answer_with_labels(prompt, chosen)

    # 将长度信息添加到输出
    return dicts_chosen

data = data.map(
        generate_and_tokenize_prompt,
        num_proc=12,
        remove_columns=column_names,
        desc="Tokenizing and processing prompts",
    )

recipe = [
    AWQModifier(
        ignore=["lm_head"],
        scheme="W4A16",
        targets=["Linear"],
    ),
]
model = AutoModelForCausalLM.from_pretrained(
    model_path, device_map="auto", torch_dtype="auto"
)

# Apply algorithms.
oneshot(
    model=model,
    dataset=data,
    recipe=recipe,
    max_seq_length=MAX_SEQUENCE_LENGTH,
    #n_parallel_calib_samples=N_PARALLEL_SAMPLES,
    num_calibration_samples=NUM_CALIBRATION_SAMPLES,
    #save_compressed=True,
    #output_dir=SAVE_DIR
)
model.save_pretrained(SAVE_DIR, save_compressed=True)
tokenizer.save_pretrained(SAVE_DIR)

print("\n\n")
print("========== SAMPLE GENERATION ==============")
input_ids = tokenizer("Hello my name is", return_tensors="pt").input_ids.to("cuda")
output = model.generate(input_ids, max_new_tokens=100)
print(tokenizer.decode(output[0]))
print("==========================================\n\n")

CUDA_VISIBLE_DEVICES=7 python /data/jcxy/haolu/workspace/training/alignment-handbook/scripts/llm_compressor.py
[2025-06-10 17:25:09,556] [INFO] [real_accelerator.py:239:get_accelerator] Setting ds_accelerator to cuda (auto detect)
INFO 06-10 17:25:12 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 06-10 17:25:12 [__init__.py:239] Automatically detected platform cuda.
/data2/jcxy/llm_model/Qwen3-0.6B
/data2/jcxy/llm_model/Qwen3-0.6B-awq-test
/data/jcxy/haolu/workspace/store/train_data/ins_data/PsyLLM_InstructV6_250529_NR_DPO.jsonl
Tokenizing: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 2200/2200 [00:12<00:00, 178.43 examples/s]
2025-06-10T17:25:29.884927+0800 | reset | INFO - Compression lifecycle reset
2025-06-10T17:25:29.887408+0800 | from_modifiers | INFO - Creating recipe from modifiers
2025-06-10T17:25:31.144693+0800 | _set_resolved_mappings | INFO - Excluded 28 from resolved mappings due to shape mismatch
2025-06-10T17:25:31.482366+0800 | _calibrate | INFO - Running AWQModifier calibration with 256 samples...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [00:30<00:00,  8.53it/s]
2025-06-10T17:26:07.580737+0800 | _apply_smoothing | INFO - Smoothing activation scales...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 84/84 [01:02<00:00,  1.34it/s]
manager stage: Modifiers initialized
2025-06-10T17:27:10.174793+0800 | initialize | INFO - Compression lifecycle initialized for 1 modifiers
manager stage: Modifiers finalized
2025-06-10T17:27:11.615433+0800 | finalize | INFO - Compression lifecycle finalized for 1 modifiers
2025-06-10T17:27:11.615620+0800 | post_process | WARNING - Optimized model is not saved. To save, please provide`output_dir` as input arg.Ex. `oneshot(..., output_dir=...)`
2025-06-10T17:27:11.658799+0800 | save_pretrained_wrapper | INFO - Fetching state_dict - this may take some time
2025-06-10T17:27:13.475187+0800 | save_pretrained_wrapper | INFO - Fetching compressor
2025-06-10T17:27:13.475651+0800 | get_model_compressor | INFO - skip_sparsity_compression_stats set to True. Skipping sparsity compression statistic calculations. No sparsity compressor will be applied.

========== SAMPLE GENERATION ==============
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Hello my name is Sarah. I am a 28-year-old graduate student in a science research lab. I am working on a project about renewable energy, specifically solar energy. I am looking for a mentorship in this field. I need help finding someone who can provide guidance, support, and a positive environment. I want to know how to connect with people in my field and what kind of resources I can use to build a network in this area. Also, I want to know if there are any specific steps
==========================================

/data/jcxy/haolu/workspace/training/alignment-handbook/scripts$ du -sh /data2/jcxy/llm_model/Qwen3-0.6B-awq-test/
1.5G    /data2/jcxy/llm_model/Qwen3-0.6B-awq-test/
(haolu) (base) haolu@ubuntu:/data/jcxy/haolu/workspace/training/alignment-handbook/scripts$ du -sh /data2/jcxy/llm_model/Qwen3-0.6B
1.5G    /data2/jcxy/llm_model/Qwen3-0.6B

**Comment by Minami-su** [comment]
(haolu) (base) haolu@ubuntu:/data/jcxy/haolu/workspace/training/alignment-handbook/scripts$ CUDA_VISIBLE_DEVICES=7 python /data/jcxy/haolu/workspace/training/alignment-handbook/scripts/llm_compressor_test.py 
[2025-06-10 17:40:21,754] [INFO] [real_accelerator.py:239:get_accelerator] Setting ds_accelerator to cuda (auto detect)
README.md: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 167/167 [00:00<00:00, 1.06MB/s]
Repo card metadata block was not found. Setting CardData to empty.
val.jsonl.zst: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 471M/471M [00:45<00:00, 10.4MB/s]
Generating validation split: 100%|████████████████████████████████████████████████████████████████████| 214670/214670 [00:15<00:00, 13967.78 examples/s]
Map: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [00:00<00:00, 2196.24 examples/s]
Tokenizing: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [00:02<00:00, 91.59 examples/s]
2025-06-10T17:41:52.737467+0800 | reset | INFO - Compression lifecycle reset
2025-06-10T17:41:52.739440+0800 | from_modifiers | INFO - Creating recipe from modifiers
2025-06-10T17:41:53.987123+0800 | _set_resolved_mappings | INFO - Excluded 28 from resolved mappings due to shape mismatch
2025-06-10T17:41:54.326926+0800 | _calibrate | INFO - Running AWQModifier calibration with 256 samples...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 256/256 [00:30<00:00,  8.44it/s]
2025-06-10T17:42:29.138464+0800 | _apply_smoothing | INFO - Smoothing activation scales...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 84/84 [00:48<00:00,  1.74it/s]
manager stage: Modifiers initialized
2025-06-10T17:43:17.357267+0800 | initialize | INFO - Compression lifecycle initialized for 1 modifiers
manager stage: Modifiers finalized
2025-06-10T17:43:18.837599+0800 | finalize | INFO - Compression lifecycle finalized for 1 modifiers
2025-06-10T17:43:18.837771+0800 | post_process | WARNING - Optimized model is not saved. To save, please provide`output_dir` as input arg.Ex. `oneshot(..., output_dir=...)`



========== SAMPLE GENERATION ==============
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Hello my name is Anna, I'm a student at the university of Oslo, I'm interested in the field of music and technology. I'm looking for a summer program in Norway. Could you help me find the program that is the best fit for me?
I need to know the program's details and how to apply. Please don't use any markdown, just plain text. Please make sure to mention the program's name, location, and the dates of the program. Also, include a brief description of the program
==========================================


2025-06-10T17:43:21.479784+0800 | save_pretrained_wrapper | INFO - Fetching state_dict - this may take some time
2025-06-10T17:43:22.114466+0800 | save_pretrained_wrapper | INFO - Fetching compressor
2025-06-10T17:43:22.114744+0800 | get_model_compressor | INFO - skip_sparsity_compression_stats set to True. Skipping sparsity compression statistic calculations. No sparsity compressor will be applied.


du -sh /data2/jcxy/llm_model/Qwen3-0.6B
1.5G    /data2/jcxy/llm_model/Qwen3-0.6B
(haolu) (base) haolu@ubuntu:/data/jcxy/haolu/workspace/training/alignment-handbook/scripts$ du -sh /data/jcxy/haolu/workspace/training/alignment-handbook/scripts/Qwen3-0.6B-awq-sym
1.5G    /data/jcxy/haolu/workspace/training/alignment-handbook/scripts/Qwen3-0.6B-awq-sym

**Comment by Minami-su** [comment]
same size,same config.json.

**Comment by Minami-su** [comment]
llmcompressor-0.5.1

**Comment by brian-dellabetta** [comment]
Hi @Minami-su , can you try with llmcompressor on main? on 0.5.1, you'll have to explicitly add `QuantizationModifier`, see example [here](https://github.com/vllm-project/llm-compressor/blob/0.5.1/examples/awq/awq_one_shot.py#L30). This has been cleaned up, we will do a 0.6.0 release in the next week or so with those changes landed

**Comment by samyakshrestha** [comment]
I am having exactly the same problem. The model that is being saved after performing AWQ is NOT quantized. Is there a way to fix this? Would really appreciate the help.

**Comment by brian-dellabetta** [comment]
Hi @samyakshrestha , see [message directly above](https://github.com/vllm-project/llm-compressor/issues/1533#issuecomment-2960260001). This is because the examples on github main are for current main version, not v0.5.1. We should be releasing 0.6.0 today to Pypi

**Comment by brian-dellabetta** [comment]
Hi @Minami-su and @samyakshrestha , v0.6.0 has been released -- https://pypi.org/project/llmcompressor/

The model should now be quantized after saving, without needing to add `QuantizationModifier`. I will close this, please re-open if you encounter any issues on v0.6.0

**Comment by samyakshrestha** [comment]
@brian-dellabetta Thank you so much for the prompt reply! I was finally able to get the issue resolved! Keep up the great work
